### DISCUSSION TITLE
triggerUncaughtException: Error: tsx must be loaded with --import instead of --loader is received after upgrading to v9

### DISCUSSION THREAD

**Discussion by jurjandrei** [webdriverio/webdriverio#13664] [Q&A] [answered]
Hello,

I am trying to upgrade to v9 and getting this error:

<img width="816" alt="Screenshot 2024-09-27 at 12 21 56" src="https://github.com/user-attachments/assets/468b0a62-8f39-4a76-b8de-949a973a76a2">

My tsconfig.json looks like this:

`{
  "compilerOptions": {
    "target": "ES2017",
    "module": "ESNext",
    "moduleResolution": "node",
    "noImplicitAny": false,
    "esModuleInterop": true,
    "allowJs": false,
    "strict": false,
    "lib": [
      "ESNext",
      "dom"
    ],
    "skipLibCheck": true,
    "outDir": "dist",
    "rootDir": ".",
    "allowSyntheticDefaultImports": true,
    "types": [
      "node",
      "@wdio/globals/types",
      "@wdio/mocha-framework",
      "@wdio/appium-service"
    ],
    "noEmit": false,
    "forceConsistentCasingInFileNames": true
  },
  "include": [
    "./**/*.ts"
  ],
  "exclude": ["node_modules"],
  "ts-node": {
    "files": true
  }
}`

I have also in package.json   `"type": "module"`

Can someone guide me on how to fix it? Thanks

**Comment by christian-bromann** [comment]
@jurjandrei can you provide a minimal reproducible example?

**Reply by jurjandrei** [reply]
@christian-bromann I just updated the version of  "@wdio/cli": from 8.40.2 to 9.1.1 and then the error was received when trying to start a test.

**Reply by jurjandrei** [reply]
```
yarn test
yarn run v1.22.22
$ cross-env wdio $npm_package_config_CONFIG_PATH --suite $npm_package_config_SUITE

node:internal/modules/run_main:125
    triggerUncaughtException(
    ^
Error: tsx must be loaded with --import instead of --loader
The --loader flag was deprecated in Node v20.6.0 and v18.19.0
    at initialize (file:///Users/****/****/****/node_modules/tsx/dist/esm/index.mjs:1:919)
    at Hooks.addCustomLoader (node:internal/modules/esm/hooks:181:24)
    at Hooks.register (node:internal/modules/esm/hooks:155:16)
    at async initializeHooks (node:internal/modules/esm/utils:317:5)
    at async customizedModuleWorker (node:internal/modules/esm/worker:109:13)

Node.js v22.0.0
error Command failed with exit code 1.
```

**Comment by CJskii** [comment]
@jurjandrei I believe this issue is caused by your node version.

Update tsx to the latest version and use at least node 20

**Reply by jurjandrei** [reply]
@CJskii I am using the latest typescript version and also node 22.

**Comment by soudai-s** [comment]
Hi, I've been encountered the same issue and resolve that to use v8.40.6 of `@wdio/cli` instead of v9.1.5.

```
node -v
v22.4.0
```

**Comment by daganilsen** [comment]
Hi. I have the same issue using node 22.1.0 and @wdio/cli 9.2.1
Any news on this? 
Thanks!

Executing test with:
npm run TestX

node:internal/modules/run_main:125
    triggerUncaughtException(
    ^
Error: tsx must be loaded with --import instead of --loader
The --loader flag was deprecated in Node v20.6.0 and v18.19.0
    at initialize (file:///C:/Projects/xyz/node_modules/tsx/dist/esm/index.mjs:1:919)
    at Hooks.addCustomLoader (node:internal/modules/esm/hooks:181:24)
    at Hooks.register (node:internal/modules/esm/hooks:155:16)
    at async initializeHooks (node:internal/modules/esm/utils:317:5)
    at async customizedModuleWorker (node:internal/modules/esm/worker:109:13)

**Comment by jenskuhrjorgensen** [comment]
I think the issue lies within this line:
```
const moduleLoaderFlag = (
                (runCmd.nodeVersion('major') >= 20 && runCmd.nodeVersion('minor') >= 6) ||
                (runCmd.nodeVersion('major') === 18 && runCmd.nodeVersion('minor') >= 19)
            )
                ? '--import'
                : '--loader'
```
https://github.com/webdriverio/webdriverio/blob/06a7c8c1b5c55b6febde7d2975cde0b43f95f1fe/packages/wdio-cli/tests/commands/run.test.ts#L132

If you are running e.g. Node 21 or 22 on a minor version that is below 6, then it will set `moduleLoaderFlag` to `--loader` instead of `--import`. I think the intention with the code was that it would set it to `--import` for all versions >= 20.6, but that is not what it does. 

I bumped my Node version from 22.3 to 22.9 and now I'm no longer getting the `Error: tsx must be loaded with --import instead of --loader` error.

**Reply by christian-bromann** [reply]
Great discovery! Mind raising a PR with a fix for this?

**Reply by jenskuhrjorgensen** [reply, marked as answer]
@christian-bromann sure, here it is:
https://github.com/webdriverio/webdriverio/pull/13768

**Reply by christian-bromann** [reply]
Thanks for your investigation and resolving the problem. I hope to release wdio tomorrow.

**Reply by avinash1406** [reply]
HI @christian-bromann ,
Today, I have freshly installed WDIO on my machine. However, when I try to run the tests, I receive the following error. It seems this issue is still not resolved. What can be done locally to fix this problem? Please let us know.

![image](https://github.com/user-attachments/assets/8d9282c3-825e-42dd-88d2-6ed350d77c33)

**Reply by christian-bromann** [reply]
> What can be done locally to fix this problem?

Can you update to the latest Node.js v20.x version?

**Reply by avinash1406** [reply]
It's having updated version 23

Do I have to downgrade it to 20 ?

**Reply by christian-bromann** [reply]
I think this got fixed in https://github.com/webdriverio/webdriverio/pull/13768 and was just not yet released.

**Reply by daganilsen** [reply]
Hi @christian-bromann. This bug was reported on @wdio/cli 9.2.1
The latest version is still @wdio/cli 9.2.1. 
When do you plan for a new  release with this fix in it?

**Reply by christian-bromann** [reply]
Sometime this week.

**Reply by daganilsen** [reply]
Super! Thanks :-)
